torch/distributions/transforms.py

```diff
@@ -274,17 +253,35 @@ def __eq__(self, other):
             return False
         return self.parts == other.parts
 
-    @constraints.dependent_property
+    @constraints.dependent_property(is_discrete=False)
     def domain(self):
         if not self.parts:
             return constraints.real
-        return self.parts[0].domain
-
-    @constraints.dependent_property
+        domain = self.parts[0].domain
+        # Adjust event_dim to be maximum among all parts.
+        event_dim = self.parts[-1].codomain.event_dim
+        for part in reversed(self.parts):
+            event_dim += part.domain.event_dim - part.codomain.event_dim
+            event_dim = max(event_dim, part.domain.event_dim)
+        assert event_dim >= domain.event_dim
+        if event_dim > domain.event_dim:
+            domain = constraints.independent(domain, event_dim - domain.event_dim)
+        return domain
```

Comment:
This section is tricky.

Reply:
Yeah, agree. I am proving that this is equivalent to [here](https://github.com/pyro-ppl/numpyro/pull/876/files#diff-be93e2377765c460352859d72dc154f1bedb2791dda750b5bea634e518e834d4R169) but it is not clear for now. :))

Reply:
@fehiepsi here's an example to show why shifting is required. Consider a fictional transform
```py
t = SquareMatrixToVectorTransform()
assert t.domain.event_dim == 2
assert t.codomain.event_dim == 1
```
Now consider
```py
c = ComposeTransform([t, t, t, t, DiscreteCosineTransform(), t.inv, t.inv, t.inv, t.inv])
assert c.domain.event_dim == 4
assert c.codomain.event_dim == 4
```
What `c` does is to flatten a 4-tensor into a vector, mix up all the elements, then reshape it.

Reply:
Yes, I agree that we need it. I just meant that
```python
            event_dim += part.domain.event_dim - part.codomain.event_dim
            event_dim = max(event_dim, part.domain.event_dim)
```
is equivalent to
```python
            event_dim = part.domain.event_dim + max(event_dim - part.codomain.event_dim, 0)
```

Reply:
By the way, I think you will need to adjust log_abs_det_jacobian (see the [current implementation](https://github.com/pyro-ppl/numpyro/pull/876/files#diff-be93e2377765c460352859d72dc154f1bedb2791dda750b5bea634e518e834d4R220-R234) in NumPyro) here and in TransformedDistribution.

Reply:
Done.